### PR TITLE
[Core] Enabling the possibility of adding variables from processes

### PR DESCRIPTION
**Pull request by ddiezrod**
**📝 Description**
This is more or less the same as https://github.com/KratosMultiphysics/Kratos/pull/9117 but using the registry (so it works in cpp and python)

The idea is to use the registry to obtain the variables that a process needs to be added to a model part at the very beginning before model parts are even imported. This improves a lot the modularity as you will not need to always add this variable in the solver.

I adapted apply_inlet_process from FluidDynApp to use this new capability.

### CODE REVIEW

File: kratos/processes/process.h
Comment (loumalouomega) [inline kratos/processes/process.h:194]: Doc

File: kratos/processes/process.h
Comment (loumalouomega) [inline kratos/processes/process.h:55]: Doc

**Comment by loumalouomega** [comment]
I will wait for test foor final review

**Comment by rubenzorrilla** [comment]
Great. As it involves changing a few crucial classes I'll add it for the next @KratosMultiphysics/technical-committee discussion. Nevertheless, at first glance everything makes sense for me.

File: kratos/processes/process.h
Comment (ddiezrod) [reply kratos/processes/process.h:194]: done

File: kratos/processes/process.h
Comment (ddiezrod) [reply kratos/processes/process.h:55]: this is just a definition, what documentation you want here?

Comment (loumalouomega) [reply kratos/processes/process.h:55]: For example: /// The historical variables map. Hash is a variable defined as a string, and it contains a list of variables (vector of strings)
```suggestion
    /// The historical variables map. Hash is a variable defined as a string, and it contains a list of variables (vector of strings)
```

Comment (ddiezrod) [reply kratos/processes/process.h:55]: you love comments too much... : P

**Comment by ddiezrod** [comment]
Test added.

**Review by loumalouomega — DISMISSED** [review]
Looks Okay to me

**Comment by ddiezrod** [comment]
@loumalouomega lucky me, there was already a test with exactly same name in ContactStructApp, I have changed the name there because I think it is more clear that way

**Review by loumalouomega — COMMENTED** [review]
Okay form my side

File: kratos/python_scripts/process_factory.py
Comment (RiccardoRossi) [inline kratos/python_scripts/process_factory.py:25]: i would say there is a problem with this: 

a process may need a variable to be in the modelpart already at construction time (i am afraid there were some examples of this).

for this reason one cannot use a "real" constructor of the process to get a prototype on the top of which we could get the function, without having constructed the real process.
What i am trying to say that the GetVariable should be a function similar to the Create one

I am cc-ing  @KratosMultiphysics/technical-committee since this was a thing we discussed.

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
i must block since i believe this is incorrect in the case a process uses a historical variable already in its construction process.

File: kratos/python_scripts/process_factory.py
Comment (rubenzorrilla) [reply kratos/python_scripts/process_factory.py:25]: > i would say there is a problem with this:
> 
> a process may need a variable to be in the modelpart already at construction time (i am afraid there were some examples of this).
> 
> for this reason one cannot use a "real" constructor of the process to get a prototype on the top of which we could get the function, without having constructed the real process. What i am trying to say that the GetVariable should be a function similar to the Create one
> 
> I am cc-ing @KratosMultiphysics/technical-committee since this was a thing we discussed.

Though you're probably right, it would be helpful to locate those. The point is that maybe these "special" processes can be refactored to avoid the variable requirement at construction time, something that I think is very unnatural.

Comment (ddiezrod) [reply kratos/python_scripts/process_factory.py:25]: Any advance on this, or do I need to find another solution on our side?

Comment (rubenzorrilla) [reply kratos/python_scripts/process_factory.py:25]: Unfortunately not... It is still in our todo list for the next KTC meeting. Sorry about this.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee had a discussion about this and the main question is if we should go for the static member our virtual one. Here are pros and cons:
Static:
- Prototype would be optional
- No need to implement the default constructor in all Processes
- It needs registry for dispatching the method (That can check if is with prototype or direct use of static function)
- Having the common implementation in the base class is not possible (To be done in a `template<ClassType, BaseType>` utility)
- `GetDefaultVariables` is virtual and we cannot call it. The idea is to have `Specification` as static member and `GetDefaultVariable` uses that as a default implementation

Virtual:
- More flexible in hierarchy
- No need for registry involvement
- It needs default constructor for all processes (Several processes has reference to the modepart which should be changed to pointer)

**Comment by rubenzorrilla** [comment]
Note for myself: after (re-)discussing this in the @KratosMultiphysics/technical-committee we decided to basically mimic what is in the `GetSpecifications` of the element but making it `static`.

**Comment by philbucher** [comment]
> Note for myself: after (re-)discussing this in the @KratosMultiphysics/technical-committee we decided to basically mimic what is in the `GetSpecifications` of the element but making it `static`.

static methods cannot be virtual :/

**Comment by rubenzorrilla** [comment]
> > Note for myself: after (re-)discussing this in the @KratosMultiphysics/technical-committee we decided to basically mimic what is in the `GetSpecifications` of the element but making it `static`.
> 
> static methods cannot be virtual :/

Yes... this is the price we will be paying.

**Comment by philbucher** [comment]
@rubenzorrilla I dont understand the technical solution
What you propose does not work in C++

**Comment by RiccardoRossi** [comment]
this is currently being discussed in  #12506, @ddiezrod can you please comment there?

**Comment by ddiezrod** [comment]
> this is currently being discussed in #12506, @ddiezrod can you please comment there?

ok, Ill close this

**Pull request closed without merging**
